### DISCUSSION TITLE
[@types/sinon] TS2694: no exported member 'FakeTimerInstallOpts'

### DISCUSSION THREAD

**Discussion by Ulrikop** [DefinitelyTyped/DefinitelyTyped#61374] [Issues with a @types package]
Hallo

after upgrading @types/sinon 10.0.12 to 10.0.13 I get the following error message:

> Error: node_modules/@types/sinon/index.d.ts:1447:53 - error TS2694: Namespace '"node_modules/@sinonjs/fake-timers/types/fake-timers-src"' has no exported member 'FakeTimerInstallOpts'.
>
> 1447         useFakeTimers: boolean | Partial<FakeTimers.FakeTimerInstallOpts>;

Has a breaking change been introduced in this patch release? Is it a bug? Or do I need to update another dependency as well?

Thanks

[a comment by an automated account is not shown]

**Comment by 43081j** [comment]
very strange.

@sandersn could this be because the `sinon` types don't have a `package.json` but do depend on `@sinonjs/fake-timers`? (external package).

so i guess if a consumer has an older fake-timers in their dependency tree, it'd use that one?

im not too sure how DT deals with dependencies, version constraints of them, etc.

**Reply by sandersn** [reply]
does sinon depend on another DT package that has a package.json? Otherwise I don't know how it would depend on `@sinonjs/fake-timers`.

Oh, it's possible that both packages define the same namespace/ambient module, which merges, and then the compiler reports the fake-timers-src one as the authoritative declaration (even though it's only one of several).

**Reply by 43081j** [reply]
the sinon DT types import from `@sinonjs/fake-timers` but don't specify it as a dependency. could that cause problems?

it doesn't seem to specify the same namespace

**Reply by sandersn** [reply]
yeah, the DT tests are supposed to install packages so that they only see other DT packages, plus anything in their own package.json. However, there must be some exception -- probably the containing DefinitelyTyped/node_modules directory is accessible, and DT itself must depend on `@sinonjs/fake-timers`. I think the fix is for sinon on DT to add a package.json with explicit dependency on `@sinonjs/fake-timers`.

**Reply by 43081j** [reply]
it looks like c46eb6974d2 got rid of it, we used to have it setup that way.

so basically `@sinonjs/fake-timers` _removed_ their shipped types, and moved consumers back to DT.

edit:

so it seems consumers with the older `@sinonjs/fake-timers` (with shipped types) as a dependency, causes the DT sinon types to import that rather than following the mapping in `tsconfig` (which points it at the types package).

some funkiness going on there i guess, with the tsconfig paths ([here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0852d33e37bc8fff1fe99cc328bde65e84e3e1c0/types/sinon/tsconfig.json#L18-L21))

**Comment by BertKleewein** [comment]
Aargh. I went through all the trouble to write a new issue and then clicked the "search for similar discussions" button and found this. After staring at this for 3 hours, here is the issue I was going to submit.  Maybe it can help the discussion:

---------------------

I hope this is clear enough. I'm in over my head with this, but I think I have it sorted out.

I started seeing a break in my code when this change showed up in my `node_modules`: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/61334

I _think_ this happens because I npm gives me `@sinonjs/fake-timers@7.1.2` by default and `@types/sinon-test@2.4.3` gives me typings that are tied to `fake-timers@9.x.x`.

I can fix this by manually adding `"@sinonjs/fake-timers": "^9.1.2"` to my `devDependencies`, so I'm not blocked by this, but others may run into it.

I'm tagging @alecgibson and @AviVahl because they've both poked around in this area, they know the history, and they are both clearly much smarter than I am.

# details

Using node 16.14.2 and npm 8.5.0, running `npm install` with this package.json 
```json
{
  "name": "silly-module",
  "main": "index.js",
  "typings": "index.d.ts",
  "devDependencies": {
    "@types/sinon-test": "^2.4.3",
    "sinon": "^11.1.2",
    "sinon-test": "^3.1.3",
    "typescript": "4.4.4"
  },
  "scripts": {
    "build": "tsc"
  }
}
```

I get this error when I run `npm run build`:

```
node_modules/@types/sinon/index.d.ts:1447:53 - error TS2694: Namespace '"/home/bertk/temp/fake-timers/node_modules/@sinonjs/fake-timers/types/fake-timers-src"' has no exported member 'FakeTimerInstallOpts'.

1447         useFakeTimers: boolean | Partial<FakeTimers.FakeTimerInstallOpts>;
                                                         ~~~~~~~~~~~~~~~~~~~~

node_modules/@types/sinon/index.d.ts:1569:67 - error TS2694: Namespace '"/home/bertk/temp/fake-timers/node_modules/@sinonjs/fake-timers/types/fake-timers-src"' has no exported member 'FakeTimerInstallOpts'.

1569         useFakeTimers(config?: number | Date | Partial<FakeTimers.FakeTimerInstallOpts>): SinonFakeTimers;
                                                                       ~~~~~~~~~~~~~~~~~~~~


Found 2 errors.
```

# extra files

I don't think these are important, but here are the other files for my minimal repro:

tsconfig.json:
```json
{
    "compilerOptions": {
        "outDir": "./dist",
        "target":"es2018",
        "module": "commonjs",
        "sourceMap": true,
        "declaration": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "removeComments": false
   },
    "include": [
        "./src/**/*.ts"
    ]
}
```

index.js:
```javascript
'use strict';

module.exports = {
  sillyFunction: require('./dist/empty.js').sillyFunction,
};
```

index.d.ts:
```typescript
export { sillyFunction } from './dist/silly';
```

src/silly.ts:
```typescript
export function sillyFunction() {
  console.log("This is a silly function, right?");
}
```

**Reply by AviVahl** [reply]
I believe the issue stems from the fact that some _older_ `@sinonjs/fake-timers` gets installed and has built-in .d.ts that does not match current `@types/sinonjs__fake-timers`. More info at: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58557

I just tested the latest versions of all packages (including `sinon-test`) and everything worked.

Steps to see latest working state:
```sh
mkdir sinontest
cd sinontest
npm init -y
npm i -D typescript sinon sinon-test @types/sinon @types/sinon-test
npx tsc --init --skipLibCheck false
mkdir src
touch src/index.ts
echo "import sinon from 'sinon';import sinonTest from 'sinon-test';console.log({sinon, sinonTest});" > src/index.ts
npx tsc --noEmit
```


My package.json contained:
```json
{
  ...
  "devDependencies": {
    "@types/sinon-test": "^2.4.3",
    "@types/sinon": "^10.0.13",
    "sinon": "^13.0.2",
    "sinon-test": "^3.1.3",
    "typescript": "^4.7.4"
  }
}
```

The output of `npm ls` for these packages:
```sh
[avi@fedora sinontest]$ npm ls sinon sinon-test @types/sinon @types/sinon-test
sinontest@1.0.0 /home/avi/projects/sinontest
├─┬ @types/sinon-test@2.4.3
│ └── @types/sinon@10.0.13 deduped
├── @types/sinon@10.0.13
├─┬ sinon-test@3.1.3
│ └── sinon@13.0.2 deduped
└── sinon@13.0.2
```

Both `@types/sinon` and `sinon` got deduped, so a single latest version for each.

I've noticed you use the older `"sinon": "^11.1.2"`, which could cause the older versions to get installed.
The `package.json` you've posted also doesn't have explicit `@types/sinon`, so it could be that `"@types/sinon-test` requests `sinon@*` and your package manager resolved and locked (package-lock.json/yarn.lock) it to an older version.

**Reply by 43081j** [reply]
FYI there's some discussion in the other post further up on this.

yes it seems to be some confusion around which package gets resolved, though the type mappings should've handled that unless im missing something.

the sinon types have a mapping to the DT fake-timers types but it seems that gets skipped or something and we end up resolving the actual fake-timers package (if you have it, i.e. the older one which shipped its own types).

will have to wait and see what the DT maintainers recommend i think as we can't really expect consumers to change their codebases to fix the problem

**Reply by AviVahl** [reply]
`@sinonjs/fake-timers` added the .d.ts into the package itself for several versions, and then removed it.
If anyone gets any of those old versions installed, the built-in .d.ts will override any `@types/sinonjs__fake-timers` that is installed, and that's where things break.

**Reply by 43081j** [reply]
of course, but im not sure why the tsconfig mapping makes no difference. the sinon types here are told explicitly to use the DT types for fake-timers (via tsconfig path mappings).

i guess those path mappings make no difference at build time in that case...

**Reply by AviVahl** [reply]
Built-in .d.ts takes precedence over tsconfig "paths", I think.
Rather than trying to force `@types/sinonjs__fake-timers` to be used with the older version of `@sinonjs/fake-timers`, I'd try getting the project to use the latest version.

`npm ls @sinonjs/fake-timers` will tell you which versions are installed.

**Comment by JaydenL33** [comment]
Hi All, we are able to fix the issue by reverting back to 10.0.12

Please find this discussion

https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/61536

**Reply by 43081j** [reply]
just waiting for some guidance from the DT maintainers on this. worth following the discussions above, in the other 2 posts.

i'd rather not rely on every consumer having to update their dependencies to make this work, but if thats the case we should at least bump the major version or something i think.

**Comment by udhayakumarcp** [comment]
upgrading the `sinon` to to verison 12.0.1 is fixed for me
